### PR TITLE
Post Categories Selection Fix

### LINKED ISSUE #9887: Bug: Categories have no check boxes
### Expected behavior
Create new blog post, menu>post settings>Categories. Categories have check boxes, or some way to see which are selected

### Actual behavior
I get a correctly indented list of categories, but no indication of which are currently selected

### Steps to reproduce the behavior

Installed Wordpress 5.2, added a few categories, in a nested structure.

On android: Create new blog post, menu>post settings>Categories.

New page opens, with list of Categories, all nested correctly.

But there's no check boxes, no highlighting, no way to tell which are selected.

Clicking them seems to have random effects on whether they're selected or not.

##### Tested on Hudl 2 , Android 5.1, WPAndroid 12.3.2
and
Honor 7, Android 6.0, WPAndroid 12.3.2

Also same problem with another site I have access to, running Wordpress 5.1.1

### PR DESCRIPTION
**Pull request by jd-alexander**
Fixes #9887 

## Findings

When the post is fetched from the `PostStore` it's returning a `post` that has different category types. So sometimes it returns one that has the selected categories and another that has the default. I tried to investigate what could be causing this but I haven't found the solution as yet. This could be merged to get around the UI issue for now or some more investigation could be done for this issue so that the whole thing can be fixed altogether.

## Solution
1. I simply replaced the checkmark drawables with new ones that are similar to that of a checkbox so that the component & code could remain intact.

Before | After 
--------|-------
        |       

## Testing
1. Go to a post. 
2. Go to post settings. 
3. Go to Categories and Tags and select Categories. 
4. Select multiple categories and return to the previous screen and back. 

## Reviewing 
Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] If it's feasible, I have added unit tests.

### CODE REVIEW

**Comment by jd-alexander** [comment]
To reviewers, this PR is still a draft because it hasn't addressed one of the bugs that's described in the issue. Nonetheless, I wanted some feedback about the approach of simply replacing the drawables and if you also had any idea why the `post` issue described above would occur. Let me know. 😄

**Review by theck13 — CHANGES_REQUESTED** [review]
The code works as expected.  There are a few optimizations we can make if that's within the scope of these changes.

We should follow the [Drawable Resources](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/docs/using-android-resources.md#drawable-resources) guidelines for naming convention of drawables.  That means `check_box_outline.xml` and `checkbox_blank_outline.xml` would be `ic_check_box_white_24dp.xml` and `ic_check_box_outline_white_24dp.xml`, respectively.  Even though those names aren't necessarily the most understandable and adding something like `checked` and `unchecked` may be more descriptive, we want to use the name given by the source repository so others can more easily use the name to lookup the original icon in the future.

We can update the `category_checkmark.xml` file too.  If we want to keep the existing behavior where the checked icon is always shown unless the item is not checked and not pressed, then we can simplify the file as follows:

```xml
<selector
    xmlns:android="http://schemas.android.com/apk/res/android">

    <item android:drawable="@drawable/ic_check_box_outline_white_24dp" android:state_checked="false" android:state_pressed="false" />
    <item android:drawable="@drawable/ic_check_box_white_24dp" />

</selector>
```

If we want to change the behavior so that the opposite icon is shown when the item is pressed, then we can update the file as follows:

```xml
<selector
    xmlns:android="http://schemas.android.com/apk/res/android">

    <item android:drawable="@drawable/ic_check_box_white_24dp" android:state_checked="false" android:state_pressed="true" />
    <item android:drawable="@drawable/ic_check_box_white_24dp" android:state_checked="true" android:state_pressed="false" />
    <item android:drawable="@drawable/ic_check_box_outline_white_24dp" android:state_checked="true" android:state_pressed="true" />
    <item android:drawable="@drawable/ic_check_box_outline_white_24dp" android:state_checked="false" android:state_pressed="false" />

</selector>
```

We should to ask the designers which behavior they prefer so we'll want to add the `[Status] Needs Design Review` label to this pull request.

We'll want to mimic the code style of other icons in the vector drawable files.  To do that, they would end up like this:

```xml
<vector
    xmlns:android="http://schemas.android.com/apk/res/android"
    android:height="24dp"
    android:width="24dp"
    android:viewportHeight="24.0"
    android:viewportWidth="24.0" >

    <path
        android:fillColor="@android:color/white"
        android:pathData="M19,3H5A2,2 0 0,0 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V5A2,2 0 0,0 19,3M19,5V19H5V5H19M10,17L6,13L7.41,11.58L10,14.17L16.59,7.58L18,9" >
    </path>

</vector>
```

```xml
<vector
    xmlns:android="http://schemas.android.com/apk/res/android"
    android:height="24dp"
    android:width="24dp"
    android:viewportHeight="24.0"
    android:viewportWidth="24.0" >

    <path
        android:fillColor="@android:color/white"
        android:pathData="M19,3H5C3.89,3 3,3.89 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V5C21,3.89 20.1,3 19,3M19,5V19H5V5H19Z" >
    </path>

</vector>
```

Notice that the `android:fillColor` attribute value is changed to `@android:color/white`.  That is intentional so that we can use a single drawable resource and tint the icon for each specific use case.  Unless the icon cannot be tinted (e.g. `ic_bell_blue_wordpress_48dp.xml` used for an app shortcut icon), we should use white as the fill color.  For that reason, we should keep the `android:checkMarkTint` attribute for the `CheckedTextView` view in the `categories_row` layout.  If we want to keep the current color scheme, we'll want to keep the `@color/neutral` value.  If we want to use the `colorControlNormal` attribute, we'll want to update the `categories_row.xml` file like this: 

```xml
<org.wordpress.android.widgets.CheckedLinearLayout
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools"
    android:id="@+id/categoryRow"
    android:clickable="false"
    android:layout_height="@dimen/category_row_height"
    android:layout_width="fill_parent"
    android:orientation="horizontal">

    <CheckedTextView
        android:id="@+id/categoryRowText"
        android:checkMark="@drawable/category_checkmark"
        android:checkMarkTint="?attr/colorControlNormal"
        android:gravity="center_vertical"
        android:layout_height="fill_parent"
        android:layout_width="fill_parent"
        android:orientation="vertical"
        android:textColor="?attr/wpColorText"
        android:textSize="@dimen/text_sz_large"
        tools:text="Category">
    </CheckedTextView>

</org.wordpress.android.widgets.CheckedLinearLayout>
```

That's another thing we should ask the designers.

**Comment by jd-alexander** [comment]
Thanks for the review @theck13 I really appreciate how thorough you have explained this all 😄. I will be making the suggested changes and then tagging the design team once I am done 👍

**Comment by jd-alexander** [comment]
For design review, below is a video of the way how the checkmark icon's state behaviors in reaction to direct actions. 

1. When the user presses the category the icon appears checked but becomes unchecked if a full press isn't done. (1s - 6s) 
2. When the user presses the category the icon is checked and to make it unchecked the user presses it again (9s - 18s)

Checkmark checked and pressed state video |
--------|
[Video Link](https://drive.google.com/file/d/19Irc-nM8KCtmRgK_ao6P-RpXokzMhEVN/view?usp=sharing)        |

**Comment by jd-alexander** [comment]
@theck13 do you have any thoughts about the bug under "Findings" I am gonna do another investigation to see if I can figure out what's happening but was just wondering if you have any thoughts 😄

**Comment by theck13** [comment]
I'm not sure why the `PostStore` would return different categories.  Maybe the store hasn't been  synced before the `Post` object is retrieved?  That _may_ be why the categories have the default  values sometimes.

**Comment by jd-alexander** [comment]
> I'm not sure why the PostStore would return different categories. Maybe the store hasn't been synced before the Post object is retrieved? That may be why the categories have the default values sometimes.

Interesting. Thanks. What odd is if you are on that screen you will see the categories selected and if you press back and then reopen they are unselected and if you do the same action again it gets reselected so it's like every other time the categories isn't properly propagated.

**Comment by mattmiklic** [comment]
> For design review, below is a video of the way how the checkmark icon's state behaviors in reaction to direct actions.

Thanks for the ping @jd-alexander! I think this works pretty well overall. In terms of visual styling, it would be nice if we could use the checkbox layout and appearance that's defined in the Material Design spec: https://material.io/components/selection-controls/#checkboxes

- 24px square boxes
- When unchecked, boxes have a neutral gray outline
- When checked, boxes have no outline, are filled with the primary color, and have a white checkmark
- Boxes sit to the left of their text label (rather than along the right-hand edge)
- Boxes sit on the 16dp keyline; their text labels sit on the 72dp keyline.

The link above has more detail on styling of checkboxes in a disabled, hover, focus, and pressed state.

**Comment by jd-alexander** [comment]
Hi @mattmiklic I made the changes you suggested! The `CheckBox` component in Android is utilizing the Material Design Library that Google created so all the styles and behavior are built in. Below are some screenshots of the new layout. 

Pressed & Unchecked | Pressed & Checked 
--------|-------
        | 

LTR | RTL 
--------|-------
        |

**Comment by jd-alexander** [comment]
Hey, @theck13 👋 so I recreated the layout and utilized the `CheckBox` so the drawables are no longer needed. Let me know what you think about the changes.

**Review by theck13 — CHANGES_REQUESTED** [review]
There are code style issues in the `categories_row.xml` file.  While they may not be flagged with Checkstyle, they don't follow the style of other XML layout files and there is even inconsistency within `categories_row.xml` itself.  There are no functional downsides to the code style issues, but there can be human readability issues.  Thus, it's best to follow the other XML layout files for code style consistency.

Are there any benefits to changing the layout and creating the new class `CheckedConstraintLayout` over keeping the existing `CheckedLinearLayout`?

Even though 16dp and 72dp paddings are used for the checkbox and text views, they are not vertically aligned with the back arrow and title text in the app bar.  `CheckBox`es have approximately 8dp padding built in to the view to help with things like the touch ripple, which makes them slightly off-center when compared with the back arrow above when using 16dp left/start padding.  There is 16dp of padding added to the `TextView` programmatically based on the level of the category so that you can see which categories have a parent/child relationship.  With the new layout, that causes the category text to be misaligned with the ***Select categories*** title in the app bar.  We can correct for that as shown in the ***Align*** screenshots below, but I find the lack of indentation of the checkbox for child categories to be visually confusing.  It's very easy to miss that a category is a child if you're scanning the list especially if you have more than a few categories.  An alternate solution would be to indent the entire row layout programmatically similar to the way the `TextView` is done currently.  Unfortunately, that makes the checkboxes and text misaligned with the back arrow and title in the app bar.  @mattmiklic, what do you think is best?  See the screenshots below for illustration.

File: WordPress/src/main/res/values/dimens.xml
Comment (mattmiklic) [reply WordPress/src/main/res/values/dimens.xml]: Thanks for implementing these changes @jd-alexander and thanks to @theck13 for these suggestions. I think either the "Align" or "Alternate" options shown above would work. The "Alternate" view with the indented child checkbox is a little more visually cluttered, but agree that it makes it easier to denote child categories at a glance. Let's go with that if it's technically feasible; if for any reason it's not we can use the "Align" style shown here.

Comment (jd-alexander) [reply WordPress/src/main/res/values/dimens.xml]: Thanks @mattmiklic I will go with the "Alternate" approach and let y'all know how it goes.

**Comment by jd-alexander** [comment]
Thanks again for another review @theck13 I really appreciate it 🙌 

> There are code style issues in the categories_row.xml file. While they may not be flagged with Checkstyle, they don't follow the style of other XML layout files and there is even inconsistency within categories_row.xml itself. There are no functional downsides to the code style issues, but there can be human readability issues. Thus, it's best to follow the other XML layout files for code style consistency.

Thanks! I will do another run through. Consistency is of utmost importance for codebases of this nature 😉 

> Are there any benefits to changing the layout and creating the new class CheckedConstraintLayout over keeping the existing CheckedLinearLayout?

I created the new layout due to the mention of the keyline usage by Matt and since  `Guidelines` are a good approach to replicating the keyline behavior I opted to use a `ConstraintLayout` since that's the preferred way to create layouts in Android these days. It was after creating the layout I truly realized the purpose of the `CheckedLinearLayout` for the `ListView`'s item checking functionality and decided to let created the `CheckedConstraintLayout`. I also didn't see any other usages of `CheckedLinearLayout` so it felt like a change without any side effects. 

> Even though 16dp and 72dp paddings are used for the checkbox and text views, they are not vertically aligned with the back arrow and title text in the app bar. CheckBoxes have approximately 8dp padding built in to the view to help with things like the touch ripple, which makes them slightly off-center when compared with the back arrow above when using 16dp left/start padding. There is 16dp of padding added to the TextView programmatically based on the level of the category so that you can see which categories have a parent/child relationship. 

Makes sense! Thanks for expounding on this! 

> With the new layout, that causes the category text to be misaligned with the Select categories title in the app bar. We can correct for that as shown in the Align screenshots below, but I find the lack of indentation of the checkbox for child categories to be visually confusing. It's very easy to miss that a category is a child if you're scanning the list especially if you have more than a few categories. An alternate solution would be to indent the entire row layout programmatically similar to the way the TextView is done currently. Unfortunately, that makes the checkboxes and text misaligned with the back arrow and title in the app bar. @mattmiklic, what do you think is best? See the screenshots below for illustration.

Thanks for these screenshots! For the implementation, I will be

1. Creating a `8dp` keyline for the `CheckBox` since it has another `8dp` internally.
2. Adding the `16dp` padding to the entire row when a node has a parent. 
3. Making the style changes.

**Comment by jkmassel** [comment]
We're freezing `14.2` today, so this PR is being bumped to `14.3`. If you'd like it to ship with `14.2`, please merge it into `release/14.2` and ping me – I'll be happy to cut a new beta release.

**Comment by oguzkocer** [comment]
Hey @jd-alexander! It looks like there hasn't been any updates on this PR for a while, could you have a look at it once you get a chance? If it's on hold for now, if you can also update the milestone with its intended version, that'd be great!

**Comment by jd-alexander** [comment]
> Hey @jd-alexander! It looks like there hasn't been any updates on this PR for a while, could you have a look at it once you get a chance? If it's on hold for now, if you can also update the milestone with its intended version, that'd be great!

Hey @oguzkocer ! I paused the work here since the material design/dark mode work was being done and these files would be touched. Once the material theme is merged into `develop` I will continue so I will bump the milestone further.

**Comment by oguzkocer** [comment]
Thanks for the update and bumping the release version @jd-alexander!

**Comment by jd-alexander** [comment]
Closing this for now as the UI for the categories changed and alot of the commits in this won't be needed to transfer the material design UI changes. Will keep the branch around nonetheless.

**Pull request closed without merging**
